File: sys/petscOptionsGetBool.m
function [value, found, errCode, toplevel] = petscOptionsGetBool(opts, pre, name)
% Gets the integer value for a particular option in the database.
%
%   [value, found, errCode, toplevel] = petscOptionsGetBool(opts, pre, name)
%   obtains a Boolean value in the data base. The flag found is PETSC_TRUE
%   if the attribute was found. The strings must be null-terminated.
%
% SEE ALSO: petscOptionsInsertString, petscOptionsHasName,
%           PetscOptionsGetInt, PetscOptionsGetReal, PetscOptionsGetString
%
% PETSc C interface:
%   PetscErrorCode  PetscOptionsGetBool(PetscOptions options,const char pre[],
%        const char name[],PetscBook *ivalue,PetscBool  *found)
% http://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Sys/PetscOptionsGetBool.html

%#codegen -args {PetscOptions, coder.typeof(char(0), [1,inf]), coder.typeof(char(0), [1,inf])}

errCode = int32(-1);

if ~coder.target('MATLAB')
    toplevel = nargout>3;
    if ~isempty(pre) && pre(end) && (toplevel || m2c_debug)
        m2c_error('MPETSc:petscOptionsGetBool:InputError', ...
            'Argument prefix must be a null-terminated string.')
    end
    if ~isempty(name) && name(end) && (toplevel || m2c_debug)
        m2c_error('MPETSc:petscOptionsGetBool:InputError', ...
            'Argument name must be a null-terminated character string.')
    end
    
    b_value = coder.opaque('PetscBool');
    b_flag = coder.opaque('PetscBool');
    
    errCode = coder.ceval('PetscOptionsGetBool', PetscOptions(opts), ...
        coder.rref(pre), coder.rref(name), coder.wref(b_value), coder.wref(b_flag));
    
    value = int32(0); %#ok<NASGU>
    value = coder.ceval('(int)', b_value);
    
    if nargout>1
        found = int32(0); %#ok<NASGU>
        found = coder.ceval('(int)', b_flag);
    end
    
    if errCode && (toplevel || m2c_debug)
        m2c_error('petsc:RuntimeError', 'PetscOptionsGetBool returned error code %d\n', errCode)
    end
end

end

File: sys/codegen/lib/petscOptionsGetBool/petscOptionsGetBool.c
#include "petscOptionsGetBool.h"
#include "m2c.h"
#include "petsc4m.h"

static void b_m2c_error(void);
static void c_m2c_error(int varargin_3);
static void m2c_error(void);
static void b_m2c_error(void)
{
  const char * msgid;
  const char * fmt;
  msgid = "Petsc4m:petscOptionsGetBool:InputError";
  fmt = "Argument name must be a null-terminated character string.";
  M2C_error(msgid, fmt);
}

static void c_m2c_error(int varargin_3)
{
  const char * msgid;
  const char * fmt;
  msgid = "petsc:RuntimeError";
  fmt = "PetscOptionsGetBool returned error code %d\n";
  M2C_error(msgid, fmt, varargin_3);
}

static void m2c_error(void)
{
  const char * msgid;
  const char * fmt;
  msgid = "Petsc4m:petscOptionsGetBool:InputError";
  fmt = "Argument prefix must be a null-terminated string.";
  M2C_error(msgid, fmt);
}

void petscOptionsGetBool(const emxArray_char_T *pre, const emxArray_char_T *name,
  int *value, int *found, int *errCode, boolean_T *toplevel)
{
  PetscOptions obj;
  PetscBool b_value;
  PetscBool b_flag;
  *toplevel = true;
  if ((pre->size[1] != 0) && (pre->data[pre->size[1] - 1] != '\x00')) {
    m2c_error();
  }

  if ((name->size[1] != 0) && (name->data[name->size[1] - 1] != '\x00')) {
    b_m2c_error();
  }

  obj = NULL;
  *errCode = PetscOptionsGetBool(obj, &pre->data[0], &name->data[0], &b_value,
    &b_flag);
  *value = (int)(b_value);
  *found = (int)(b_flag);
  if (*errCode != 0) {
    c_m2c_error(*errCode);
  }
}

void petscOptionsGetBool_initialize(void)
{
}

void petscOptionsGetBool_terminate(void)
{
}
